Storage/SQL/SQLiteBookmarksModel.swift

```diff
@@ -280,7 +275,27 @@ extension SQLiteBookmarks {
             "ORDER BY title ASC",
             ].joinWithSeparator(" ")
 
-        return self.db.runQuery(withIcon, args: args, factory: BookmarkFactory.factory)
+        let sql = includeIcon ? withIcon : values + " ORDER BY title ASC"
+        return isEditableForDirection(direction) >>== { editable in
+            args.insert((editable ? 1 : 0) as AnyObject, atIndex: 0)
+            return self.db.runQuery(sql, args: args, factory: BookmarkFactory.factory)
+        }
+    }
+
+    private func isBufferEmpty() -> Deferred<Maybe<Bool>> {
+        return self.db.queryReturnsNoResults("SELECT 1 FROM \(TableBookmarksBuffer)")
+    }
+
+    private func isMirrorEmpty() -> Deferred<Maybe<Bool>> {
+        return self.db.queryReturnsNoResults("SELECT 1 FROM \(TableBookmarksMirror)")
+    }
+
+    func isEditableForDirection(direction: Direction) -> Deferred<Maybe<Bool>> {
+        return accumulate([isBufferEmpty, isMirrorEmpty]) >>== { checks in
+            let bufferEmpty = checks[0]
+            let mirrorEmpty = checks[1]
+            return deferMaybe(bufferEmpty && direction == .Local ? true : mirrorEmpty)
```

Comment:
Even better, use `exists`:

```
SELECT exists(SELECT 1 FROM \(TableBookmarksBuffer)) AS hasBuffer, …
```

which *might* be smart enough that you can insert it into your `isEditable` clause above, and skip this whole function.